util-images/network/netperfbenchmark/pkg/worker/worker.go

```diff
@@ -0,0 +1,287 @@
+}
+
+// StartTCPServer starts iperf server for tcp measurements
+func (w *Worker) StartTCPServer(rw http.ResponseWriter, req *http.Request) {
+	klog.Info("Starting TCP Server")
+	w.startWork(rw, req, ProtocolTCP, "iperf", tcpServerArguments)
+}
+
+// StartUDPServer starts iperf server for udp measurements
+func (w *Worker) StartUDPServer(rw http.ResponseWriter, req *http.Request) {
+	klog.Info("Starting UDP Server")
+	w.startWork(rw, req, ProtocolUDP, "iperf", udpServerArguments)
+}
+
+// StartHTTPClient starts an http client for http measurements
+func (w *Worker) StartHTTPClient(rw http.ResponseWriter, req *http.Request) {
+	klog.Info("Starting HTTP Client")
+	w.startWork(rw, req, ProtocolHTTP, "siege", httpClientCmd)
+}
+
+// StartTCPClient starts iperf client for tcp measurements
+func (w *Worker) StartTCPClient(rw http.ResponseWriter, req *http.Request) {
+	klog.Info("Starting TCP Client")
+	w.startWork(rw, req, ProtocolTCP, "iperf", tcpClientArguments)
+}
+
+// StartUDPClient starts iperf client for udp measurements
+func (w *Worker) StartUDPClient(rw http.ResponseWriter, req *http.Request) {
+	klog.Info("Starting UDP Client")
+	w.startWork(rw, req, ProtocolUDP, "iperf", udpClientArguments)
+}
+
+// Handler handles http requests for http measurements
+func (w *Worker) Handler(rw http.ResponseWriter, req *http.Request) {
+	w.sendResponse("hi", rw, http.StatusOK)
+}
+
+func (w *Worker) startWork(rw http.ResponseWriter, req *http.Request, protocol string, cmd string, args []string) {
+	// Below loop populates template parameters with actual values from the http request object
+	for i, arg := range args {
+		if idx := strings.Index(arg, "{"); idx != -1 {
+			param := arg[strings.Index(arg, "{")+1 : strings.Index(arg, "}")]
+			klog.Infof("arg with template: %v", param)
+			valMap, err := GetValuesFromURL(req, []string{param})
+			if err != nil {
+				w.sendResponse(startWorkResponse{error: err.Error()}, rw, http.StatusBadRequest)
+				return
+			}
+			args[i] = strings.Replace(arg, "{"+param+"}", valMap[param], 1)
+			klog.Infof("Value after resolving template: %v", args[i])
+		}
+	}
+	valMap, err := GetValuesFromURL(req, []string{"timestamp", "duration"})
+	if err != nil {
+		w.sendResponse(startWorkResponse{error: err.Error()}, rw, http.StatusBadRequest)
+		return
+	}
+	tsint, err := strconv.ParseInt(valMap["timestamp"], 10, 64)
+	if err != nil {
+		klog.Infof("Invalid timestamp: %v %v", valMap["timestamp"], err)
+	}
+	go w.schedule(protocol, tsint, valMap["duration"], cmd, args)
+	w.sendResponse(startWorkResponse{}, rw, http.StatusOK)
+	return
+}
+
+func (w *Worker) schedule(protocol string, futureTimestamp int64, duration string, command string, args []string) {
+	//If future time is in past,run immediately
+	klog.Infof("About to wait for futuretime: %v", futureTimestamp)
+	klog.Infof("Current time: %v", time.Now().Unix())
+	time.Sleep(time.Duration(futureTimestamp-time.Now().Unix()) * time.Second)
+	w.startedAt = time.Now()
```

Comment:
startedAt would be removed after few initial runs in gcs, it is introduced just for checking the delay and tuning accordingly.

Reply:
As mentioned above we want the workers to start measurement at almost the same time(hence introduced future time, which would be same for all workers and all workers would start measuring at that time).To know the proper future time to be set(ie offset in seconds from current time) especially in case of huge number of pods, we thought of printing the time difference between future time given to a pod and the actual time it started. A negative difference would mean the pod itself got the message to start the measurement after future time had lapsed and the value would help us tune the offset to be given.

Reply:
OK - so "negative difference" is what you would are interested in catching.

So I'm not sure it's the best way of passing that information.
What I suggest doing is to:

- extending metricResponse with "workerDelay" (or sth like)
  [that basically would signal if the worker was started too late]
- removing the current "WorkerStartAt" field (which has pretty strange format)
- in here just recording the delay (if futureTime is already in the past)
  [so basically storing "workerDelay" instead of "startedAt" timestamp

Reply:
Modified as suggested.